cli/tests/cli_test.go

```diff
@@ -0,0 +1,245 @@
+				c.destroy(t, dirTestdata)
+			}()
+		}
+
+		for _, testcase := range c.testers {
+			t.Run(fmt.Sprintf("command=%s,test=%s", c.name, testcase.name), func(t *testing.T) {
+				testcase.tester(t, dirTestdata)
+			})
+		}
+	}
+}
+
+func testValidatePassing(t *testing.T, dirTestdata string) {
+	stdOut, stdErr := runCft(t, false, cwd, []string{
+		"validate",
+		"my-networks",
+		"--policy-path",
+		filepath.Join(dirTestdata, "policies", "test_all"),
+	})
+	checkOutput(t, append(stdOut, stdErr...), []string{"No violations found."})
+
+	stdOut, stdErr = runCft(t, false, cwd, []string{
+		"validate",
+		"my-firewalls",
+		"--policy-path",
+		filepath.Join(dirTestdata, "policies", "test_all"),
+	})
+	checkOutput(t, append(stdOut, stdErr...), []string{"No violations found."})
+
+	stdOut, stdErr = runCft(t, false, cwd, []string{
+		"validate",
+		"my-instance-prod-1",
+		"--policy-path",
+		filepath.Join(dirTestdata, "policies", "test_all"),
+	})
+	checkOutput(t, append(stdOut, stdErr...), []string{"No violations found."})
+
+	stdOut, stdErr = runCft(t, false, cwd, []string{
+		"validate",
+		"project-iam",
+		"--policy-path",
+		filepath.Join(dirTestdata, "policies", "test_all"),
+	})
+	checkOutput(t, append(stdOut, stdErr...), []string{"No violations found."})
+}
+
+func testValidateFailing(t *testing.T, dirTestdata string) {
+	stdOut, stdErr := runCft(t, false, cwd, []string{
+		"validate",
+		"my-networks",
+		"--policy-path",
+		filepath.Join(dirTestdata, "policies", "test_none"),
+	})
+	checkOutput(t, append(stdOut, stdErr...), []string{"No violations found."})
+
+	stdOut, stdErr = runCft(t, false, cwd, []string{
+		"validate",
+		"my-firewalls",
+		"--policy-path",
+		filepath.Join(dirTestdata, "policies", "test_none"),
+	})
+	checkOutput(t, append(stdOut, stdErr...), []string{
+		"Found Violations",
+		"Constraint restrict-firewall-rule-world_open",
+		"//compute\\.googleapis\\.com/projects/.+/global/firewalls/allow-proxy-from-inside-dev",
+		"//compute\\.googleapis\\.com/projects/.+/global/firewalls/allow-proxy-from-inside-prod",
+	})
+
+	stdOut, stdErr = runCft(t, false, cwd, []string{
+		"validate",
+		"my-instance-prod-1",
+		"--policy-path",
+		filepath.Join(dirTestdata, "policies", "test_none"),
+	})
+	checkOutput(t, append(stdOut, stdErr...), []string{
+		"Found Violations",
+		"Constraint gcp-compute-zone",
+		"//compute\\.googleapis\\.com/projects/.+/zones/us-central1-a/instances/my-instance-prod-1",
+	})
+
+	stdOut, stdErr = runCft(t, false, cwd, []string{
+		"validate",
+		"project-iam",
+		"--policy-path",
+		filepath.Join(dirTestdata, "policies", "test_none"),
+	})
+	checkOutput(t, append(stdOut, stdErr...), []string{
+		"Found Violations",
+		"Constraint iam_ban_roles",
+		"//cloudresourcemanager\\.googleapis\\.com/projects/12345",
+	})
+}
+
+func setupValidate(t *testing.T, dirTestdata string) {
+	runCft(t, false, cwd, []string{"apply", filepath.Join(dirTestdata, "deployment")})
```

Comment:
Validation tests should ideally not require creating *any* actual live resources. (It doesn't for TFV.)

Reply:
I don't think it is possible. Terraform is different from DM